### ISSUE TITLE
Issue with database id when using multiple plp viewer

### ISSUE THREAD

**Issue by ahijazy** [OHDSI/PatientLevelPrediction#336]
**Describe the bug**
Getting an error message:
Error in if (databaseId %in% names(databaseList)) {: argument is of length zero
Traceback:
1. runMultiplePlp(databaseDetails = db_det_erspc, modelDesignList = list(modelDesign1, 
 .     modelDesign2, modelDesign3), onlyFetchData = F, logSettings = createLogSettings(), 
 .     saveDirectory = "./results")
2. insertResultsToSqlite(resultLocation = saveDirectory, cohortDefinitions = cohortDefinitions, 
 .     databaseList = createDatabaseList(cdmDatabaseSchemas = databaseDetails$cohortDatabaseSchema, 
 .         cdmDatabaseNames = databaseDetails$cdmDatabaseName, databaseRefIds = databaseDetails$cdmDatabaseId), 
 .     sqliteLocation = sqliteLocation)
3. addMultipleRunPlpToDatabase(conn = conn, databaseSchemaSettings = createDatabaseSchemaSettings(resultSchema = "main"), 
 .     cohortDefinitions = cohortDefinitions, databaseList = databaseList, 
 .     resultLocation = resultLocation, modelSaveLocation = sqliteLocation)
4. addRunPlpToDatabase(runPlp = runPlp, conn = conn, databaseSchemaSettings = databaseSchemaSettings, 
 .     cohortDefinitions = cohortDefinitions, databaseList = databaseList, 
 .     modelSaveLocation = modelSaveLocation)
5. addDatabase(conn = conn, databaseSchemaSettings = databaseSchemaSettings, 
 .     databaseList = databaseList, databaseSchema = developmentDatabase, 
 .     databaseId = developmentDatabaseRefId)

 when running the runMultiplePlp function, the code used to work fine before the last update, the analysis seems to run fine but I also cannot use the shiny viewer, I get the error message:

Connecting using SQLite driver
Warning message:
“Error in : Could not connect to database:
unable to open database file”

**Set up (please run in R "sessionInfo()" and copy the output here):**
R version 4.1.1 (2021-08-10)
Platform: x86_64-conda-linux-gnu (64-bit)
Running under: Ubuntu 20.04.3 LTS

Matrix products: default
BLAS/LAPACK: /opt/conda/lib/libopenblasp-r0.3.18.so

locale:
 [1] LC_CTYPE=en_US.UTF-8          LC_NUMERIC=C                 
 [3] LC_TIME=en_US.UTF-8           LC_COLLATE=en_US.UTF-8       
 [5] LC_MONETARY=en_US.UTF-8       LC_MESSAGES=en_US.UTF-8      
 [7] LC_PAPER=en_US.UTF-8          LC_NAME=en_US.UTF-8          
 [9] LC_ADDRESS=en_US.UTF-8        LC_TELEPHONE=en_US.UTF-8     
[11] LC_MEASUREMENT=en_US.UTF-8    LC_IDENTIFICATION=en_US.UTF-8

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] shiny_1.7.2                  PatientLevelPrediction_6.0.0
[3] FeatureExtraction_3.2.0      Andromeda_0.6.1             
[5] dplyr_1.0.10                 DatabaseConnector_5.1.0     
[7] devtools_2.4.3               usethis_2.1.5               

loaded via a namespace (and not attached):
  [1] fs_1.5.2                reactable_0.3.0         fontawesome_0.3.0      
  [4] bit64_4.0.5             httr_1.4.4              PRROC_1.3.1            
  [7] rprojroot_2.0.3         repr_1.1.3              tools_4.1.1            
 [10] backports_1.4.1         bslib_0.4.0             DT_0.25                
 [13] utf8_1.2.2              R6_2.5.1                lazyeval_0.2.2         
 [16] colorspace_2.0-3        DBI_1.1.3               withr_2.5.0            
 [19] tidyselect_1.1.2        prettyunits_1.1.1       processx_3.5.2         
 [22] bit_4.0.4               compiler_4.1.1          Cyclops_3.2.1          
 [25] cli_3.4.1               plotly_4.10.0           desc_1.4.0             
 [28] triebeard_0.3.0         sass_0.4.2              scales_1.2.1           
 [31] checkmate_2.1.0         callr_3.7.0             pbdZMQ_0.3-6           
 [34] digest_0.6.29           base64enc_0.1-3         pkgconfig_2.0.3        
 [37] htmltools_0.5.3         sessioninfo_1.2.2       dbplyr_2.2.1           
 [40] fastmap_1.1.0           htmlwidgets_1.5.4       rlang_1.0.6            
 [43] RSQLite_2.2.17          jquerylib_0.1.4         generics_0.1.3         
 [46] jsonlite_1.8.2          crosstalk_1.2.0         zip_2.2.1              
 [49] magrittr_2.0.3          Matrix_1.4-0            munsell_0.5.0          
 [52] Rcpp_1.0.9              IRkernel_1.2            fansi_1.0.3            
 [55] shinycssloaders_1.0.0   lifecycle_1.0.2         yaml_2.3.5             
 [58] pROC_1.18.0             pkgbuild_1.3.0          plyr_1.8.7             
 [61] grid_4.1.1              blob_1.2.3              promises_1.2.0.1       
 [64] shinydashboard_0.7.2    crayon_1.5.2            lattice_0.20-45        
 [67] IRdisplay_1.0           splines_4.1.1           hms_1.1.2              
 [70] ps_1.6.0                pillar_1.8.1            uuid_1.0-3             
 [73] ParallelLogger_3.0.1    markdown_1.1            xgboost_1.6.0.1        
 [76] pkgload_1.2.4           glue_1.6.2              evaluate_0.16          
 [79] getPass_0.2-2           data.table_1.14.2       remotes_2.4.2          
 [82] memuse_4.2-1            vctrs_0.4.2             httpuv_1.6.6           
 [85] urltools_1.7.3          testthat_3.1.1          gtable_0.3.1           
 [88] purrr_0.3.4             tidyr_1.2.1             reactR_0.4.4           
 [91] assertthat_0.2.1        ggplot2_3.3.6           cachem_1.0.6           
 [94] xfun_0.33               mime_0.12               OhdsiShinyModules_1.0.0
 [97] xtable_1.8-4            later_1.3.0             viridisLite_0.4.1      
[100] survival_3.2-13         tibble_3.1.8            rJava_1.0-6            
[103] SqlRender_1.10.0        memoise_2.0.1           ellipsis_0.3.2         
**To Reproduce**
Enter the code you tried to run here (with sensitive information such as passwords removed)

**Comment by jreps** [comment]
Sorry, I think this may be due to a slight change to the databaseDetails in the update.  I think if you are using an older version of the databaseDetails it may cause the issue. Are you creating db_det_erspc using createDatabaseDetails() ? I should have made the change in databaseDetails more clear when I edited the code - I'll add a check to warn that the input has changed if the old input is used. 

The error is in happening when inserting the results into the sqlite database, the shiny needs the sqlite database to run, so once we get the code to run without error, the shiny will also work.

**Comment by ahijazy** [comment]
Thank you for your swift reply. Yes db_det_erspc was created using createDatabaseDetails(). Could you please send me a link or guide me on how to define new database details?

**Comment by Albert-Saporta** [comment]
I also have the same issues regarding the shinny viewer and defining new database details.

**Comment by jreps** [comment]
Sorry for the slow reply, was away at OHDSI and then a 2 day work meeting.  

The error suggests that a NULL is creeping in for the cdmDatabaseId, playing with Eunomia, I found I get the error if I set cdmDatabaseId = NULL in createDatabaseDetails

databaseDetails <- createDatabaseDetails(
  connectionDetails = connectionDetails, 
  cdmDatabaseSchema = "main", 
  cdmDatabaseName = "dfsfew2", 
  cdmDatabaseId = NULL, 
  cohortDatabaseSchema = "main", 
  cohortTable = "cohort", 
  outcomeDatabaseSchema = "main", 
  outcomeTable =  "cohort",
  targetId = 1, 
  outcomeIds = 3, #make this ids
  cdmVersion = 5)

I had code to cast a NULL to '' but that does not overcome the NULL issue.  I'll update the createDatabaseDetails to return an error when NULL is input instead, but for now you can get past the error by making sure you enter some string for cdmDatabaseId =  'a unique identifier for your database' such as cdmDatabaseId = 'Eunomia v1'

**Comment by ahijazy** [comment]
Great, thanks 👍

**Comment by CVilinsky** [comment]
@jreps @ahijazy 
Hi!
We also got the same error, but we did configure the cdmDatabaseId.

Here is our code:
```
databaseDetails <- PatientLevelPrediction::createDatabaseDetails(
  connectionDetails = connectionDetails, 
  cdmDatabaseSchema = cdmDatabaseSchema, 
  cdmDatabaseName = cdmDatabaseName, 
  cdmDatabaseId = 'omop_demo_v1',
  cohortDatabaseSchema = cohortDatabaseSchema, 
  cohortTable = cohortTable, 
  tempEmulationSchema=tempEmulationSchema,
  outcomeDatabaseSchema = cohortDatabaseSchema,  
  outcomeTable = cohortTable, 
  targetId=22,
  outcomeIds = 19,
  cdmVersion=5
)

modelDesign1 <- PatientLevelPrediction::createModelDesign(
  targetId = 22, 
  outcomeId = 19, 
  covariateSettings = FeatureExtraction::createCovariateSettings(
    useDemographicsAgeGroup = T,
    useDemographicsGender = T,
    useConditionGroupEraLongTerm = T
  ),
  restrictPlpDataSettings = PatientLevelPrediction::createRestrictPlpDataSettings(
    sampleSize = 100000
  ),
  splitSettings = PatientLevelPrediction::createDefaultSplitSetting(),
  populationSettings = PatientLevelPrediction::createStudyPopulationSettings(), 
  featureEngineeringSettings = PatientLevelPrediction::createFeatureEngineeringSettings(),
  sampleSettings =  PatientLevelPrediction::createSampleSettings(),
  preprocessSettings = PatientLevelPrediction::createPreprocessSettings(), 
  modelSettings = PatientLevelPrediction::setAdaBoost(), 
  runCovariateSummary = T
)

modelDesign2 <- PatientLevelPrediction::createModelDesign(
  targetId = 22, 
  outcomeId = 19, 
  covariateSettings = FeatureExtraction::createCovariateSettings(
    useDemographicsAgeGroup = T,
    useDemographicsGender = T,
    useConditionGroupEraLongTerm = T
  ),
  restrictPlpDataSettings = PatientLevelPrediction::createRestrictPlpDataSettings(
    sampleSize = 100000
  ),
  splitSettings = PatientLevelPrediction::createDefaultSplitSetting(),
  populationSettings = PatientLevelPrediction::createStudyPopulationSettings(), 
  featureEngineeringSettings = PatientLevelPrediction::createFeatureEngineeringSettings(),
  preprocessSettings = PatientLevelPrediction::createPreprocessSettings(),
  sampleSettings =  PatientLevelPrediction::createSampleSettings(),
  modelSettings = PatientLevelPrediction::setLassoLogisticRegression(), 
  runCovariateSummary = T
)

modelDesign3 <- PatientLevelPrediction::createModelDesign(
  targetId = 22, 
  outcomeId = 19, 
  covariateSettings = FeatureExtraction::createCovariateSettings(
    useDemographicsAgeGroup = T,
    useDemographicsGender = T,
    useConditionGroupEraLongTerm = T
  ),
  restrictPlpDataSettings = PatientLevelPrediction::createRestrictPlpDataSettings(
    sampleSize = 100000
  ),
  splitSettings = PatientLevelPrediction::createDefaultSplitSetting(),
  populationSettings = PatientLevelPrediction::createStudyPopulationSettings(), 
  featureEngineeringSettings = PatientLevelPrediction::createFeatureEngineeringSettings(),
  preprocessSettings = PatientLevelPrediction::createPreprocessSettings(),
  sampleSettings =  PatientLevelPrediction::createSampleSettings(),
  modelSettings = PatientLevelPrediction::setGradientBoostingMachine(),
  runCovariateSummary = T
)

modelDesignList <- list(
  modelDesign1,
  modelDesign2,
  modelDesign3
)

PatientLevelPrediction::runMultiplePlp(
  databaseDetails = databaseDetails,
  modelDesignList = modelDesignList,
  saveDirectory = file.path(outputFolder,'multiple_plp')
)
```

The version of PatientLevelPrediction we are using is 6.3.1 .

**Comment by jreps** [comment]
HI @CVilinsky - thanks for letting me know about this issue. Do you get exactly the same error message?  Is it possible to copy the log here?  Is your outputFolder a relative or absolute path?

**Comment by CVilinsky** [comment]
@jreps Thank you for the quick reply!
outputFolder is a absolute path, the files are created there (and when I used the runPlp function, everything went well).

The error is identical:
`Error in if (databaseId %in% names(databaseList)) { : 
  argument is of length zero`

But while looking at the log before I sent it, I saw that I had the error `Error: ModuleNotFoundError: No module named 'scipy'`
So I removed the specific model that tried to use it, and now it works!
